src/levanter/models/diva.py

```diff
@@ -0,0 +1,342 @@
+
+    def resize_vocab(self, new_size: int, key: Optional[PRNGKeyArray] = None) -> "DivaModel":
+        new_decoder = self.decoder.resize_vocab(new_size, key)
+        return dataclasses.replace(self, decoder=new_decoder)
+
+    @classmethod
+    def init(
+        cls,
+        Vocab: Axis,
+        config: DivaConfig,
+        *,
+        key,
+        init_from_submodels: bool = False,
+    ) -> "DivaModel":
+        k_query, k_projection, k_enc, k_connector, k_dec = maybe_rng_split(key, 5)
+
+        query_tokens = hax.random.normal(k_query, (config.Pos, config.enc_config.Embed)) * 0.02
+        projection = hnn.Linear.init(
+            In=config.enc_config.Embed.alias("whisp_embed"), Out=config.dec_config.Embed, key=key
+        )
+
+        if init_from_submodels:
+            llm: Union[LlamaLMHeadModel | MistralLMHeadModel | GemmaLMHeadModel] = HFCheckpointConverter(
+                type(config.dec_config), config.reference_decoder
+            ).load_pretrained(
+                config.dec_config.model_type,
+                config.reference_decoder,
+                config.dec_config,
+            )  # type: ignore[assignment]
+            whisper: WhisperModel = HFCheckpointConverter(
+                WhisperConfig, config.reference_encoder, ignore_prefix="model"
+            ).load_pretrained(
+                WhisperModel,
+                config.reference_encoder,
+                config.enc_config,
+            )  # type: ignore[assignment]
+            encoder = whisper.encoder
+            connector = whisper.decoder
+            decoder = llm
+        else:
+            encoder = WhisperEncoder.init(config.enc_config, key=k_enc)
+            connector = WhisperDecoder.init(config.enc_config, key=k_connector)
+            decoder = config.dec_config.model_type.init(Vocab, config.dec_config, key=k_dec)
+
+        return cls(query_tokens, projection, encoder, connector, decoder, config)
+
+    @property
+    def query_position_embeds(self) -> NamedArray:
+        return self.connector.embeddings.position_embeddings.embed(hax.arange(self.config.Pos))
+
+    def __call__(
+        self,
+        mel: NamedArray,
+        input_ids: NamedArray,
+        attn_mask: Optional[AttentionMask | NamedArray] = None,
+        pad_token_id: int = 128002,
+        *,
+        key=None,
+    ) -> NamedArray:
+        # Setup
+        Batch = input_ids.resolve_axis("batch")
+        OtherAxes = hax.axis.eliminate_axes(input_ids.axes, "position")
+        causal_mask = AttentionMask.causal()
+        if attn_mask is not None:
+            causal_mask = causal_mask & attn_mask
+        k_encoder, k_connector, k_decoder, k_head = maybe_rng_split(key, 4)
+
+        # Encode Audio With Whisper Encoder
+        audio_features = self.encoder(mel, key=k_encoder)
+
+        # Convert to Virtual LLM Tokens
+        virt_whisper_tokens = self.connector.transformer(
+            (self.query_tokens + self.query_position_embeds).broadcast_axis(OtherAxes),
```

Comment:
hrm i wouldn't think this should be necessary

Reply:
I'll double check. I've somewhat forgotten why I added these explicit broadcasts.